Core/src/Core/Filters/Algorithms/CalculateArrayHistogram.cpp

```diff
@@ -0,0 +1,232 @@
+#include "CalculateArrayHistogram.hpp"
+
+#include "Core/Filters/CalculateArrayHistogramFilter.hpp"
+
+#include "complex/Common/Range.hpp"
+#include "complex/DataStructure/DataArray.hpp"
+#include "complex/DataStructure/DataGroup.hpp"
+#include "complex/Utilities/ParallelDataAlgorithm.hpp"
+
+#include <algorithm>
+#include <chrono>
+#include <tuple>
+
+using namespace complex;
+
+namespace
+{
+template <typename T>
+class GenerateHistogramFromData
+{
+public:
+  GenerateHistogramFromData(CalculateArrayHistogram& filter, const int32 numBins, const DataArray<T>& inputArray, Float64Array& histogram, std::atomic<usize>& overflow,
+                            std::tuple<bool, float64, float64>& range, size_t progIncrement)
+  : m_Filter(filter)
+  , m_NumBins(numBins)
+  , m_InputArray(inputArray)
+  , m_Histogram(histogram)
+  , m_Overflow(overflow)
+  , m_Range(range)
+  , m_ProgIncrement(progIncrement)
+  {
+  }
+  ~GenerateHistogramFromData() = default;
+
+  void generate(const usize start, const usize end) const
+  {
+    // tuple visualization: Histogram = {(bin maximum, count), (bin maximum, count), ... }
+    float64 min = std::numeric_limits<float>::max();
+    float64 max = -1.0 * std::numeric_limits<float>::max();
+    if(std::get<0>(m_Range))
+    {
+      min = std::get<1>(m_Range);
+      max = std::get<2>(m_Range);
+    }
+    else
+    {
+      auto minMax = std::minmax_element(m_InputArray.begin(), m_InputArray.end());
+      min = (static_cast<float64>(*minMax.first) - 1);  // ensure upper limit encapsulates max value
+      max = (static_cast<float64>(*minMax.second) + 1); // ensure lower limit encapsulates min value
+    }
+
+    const float64 increment = (max - min) / static_cast<float64>(m_NumBins);
+    if(m_NumBins == 1) // if one bin, just set the first element to total number of points
+    {
+      m_Histogram[0] = max;
+      m_Histogram[1] = end;
+    }
+    else
+    {
+      size_t progCounter = 0;
+      for(usize i = start; i < end; i++)
+      {
+        if(progCounter > m_ProgIncrement)
+        {
+          m_Filter.updateThreadSafeProgress(progCounter);
+          progCounter = 0;
+        }
+        if(m_Filter.getCancel())
+        {
+          return;
+        }
+        const auto bin = std::floor((m_InputArray.at(i) - min) / increment);
```

Comment:
DataArray<T> has the [] operators which don't do bounds checking whereas the .at(i) does do bounds checking. This will slow down the algorithm for bounds checking that does not need to be performed. The loop control variables will ensure the correct bounds are followed.